sw/host/opentitantool/src/command/console.rs

```diff
@@ -0,0 +1,237 @@
+        let mut stdin = std::io::stdin();
+
+        // Set up resources specified by the command line parameters.
+        let mut console = InnerConsole {
+            logfile: self.logfile.as_ref().map(File::create).transpose()?,
+            deadline: self.timeout.map(|t| Instant::now() + Duration::new(t, 0)),
+            exit_success: self
+                .exit_success
+                .as_ref()
+                .map(|s| Regex::new(s.as_str()))
+                .transpose()?,
+            exit_failure: self
+                .exit_failure
+                .as_ref()
+                .map(|s| Regex::new(s.as_str()))
+                .transpose()?,
+            ..Default::default()
+        };
+
+        if !self.quiet {
+            println!("Starting interactive console");
+            println!("[CTRL+C] to exit.\n");
+        }
+        {
+            // Put the terminal into raw mode.  The tty guard will restore the
+            // console settings when it goes out of scope.
+            let _stdin_guard = if isatty(stdin.as_raw_fd())? {
+                let mut guard = TtyModeGuard::new(stdin.as_raw_fd())?;
+                guard.set_raw_mode()?;
+                Some(guard)
+            } else {
+                None
+            };
+            let _stdout_guard = if isatty(stdout.as_raw_fd())? {
+                let mut guard = TtyModeGuard::new(stdout.as_raw_fd())?;
+                guard.set_raw_mode()?;
+                Some(guard)
+            } else {
+                None
+            };
+            console.interact(transport, &mut stdin, &mut stdout)?;
+        }
+        if !self.quiet {
+            println!("\n\nExiting interactive console.");
+        }
+        Ok(None)
+    }
+}
+
+#[derive(Default)]
+struct InnerConsole {
+    logfile: Option<File>,
+    deadline: Option<Instant>,
+    exit_success: Option<Regex>,
+    exit_failure: Option<Regex>,
+    buffer: String,
+}
+
+enum ExitStatus {
+    None,
+    ExitSuccess,
+    ExitFailure,
+}
+
+impl InnerConsole {
+    const CTRL_C: u8 = 3;
+    const BUFFER_LEN: usize = 1024;
+
+    // Runs an interactive console until CTRL_C is received.
+    fn interact<R>(
+        &mut self,
+        transport: &mut dyn Transport,
+        stdin: &mut R,
+        stdout: &mut impl Write,
+    ) -> Result<()>
+    where
+        R: Read + AsRawFd,
+    {
+        let mut uart = transport.uart()?;
+        let mut buf = [0u8; 256];
+
+        loop {
+            if let Some(deadline) = self.deadline {
+                if Instant::now() > deadline {
+                    // If we have an exit success condition, then a timeout
+                    // should be an error.
+                    if self.exit_success.is_some() {
+                        return Err(anyhow!("Console timeout exceeded"));
+                    } else {
+                        break;
+                    }
+                }
+            }
+
+            // This loop _should_ really use `poll` to learn when the
```

Comment:
Maybe a thing to do in the future is to add a poll() function that rudely blocks when you can't find the fd?

Reply:
The `Uart` trait's `read_timeout` does exactly what you suggest: blocks on that single resource for a user-specified period of time.  The comment is lamenting the fact that the "proper" way to do things is via the unix poll (and/or friends) syscall to block on a set of resources and wake up when one of them becomes ready.

I've added a TODO to re-evaluate this once we have more backends written and can make a better decision about multi-fd blocking.